text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,153 @@
+
+Our support for various authentication methods (including 2FA, SSO, and SAML) helps mitigate access of potentially sensitive information, but API tokens cannot support these additional authentication gates.
+
+We can help customers further protect their accounts and data by providing a means of auto-detecting leaked API tokens.
+
+# Options Considered
+
+## Option #1
+
+This option strives to limit impact to the database by avoiding large bulk operations. To do this we'll distribute the complete implementation over two features. This also allows for a smooth transition of self-hosted instances keeping pace with releases.
+
+**In the first release:**
+
+1. The frontend is updated to no longer display the token value for existing tokens.
+2. The frontend is updated to accept a `name` when creating new tokens.
+3. A notification/warning in the UI should be displayed recommending users recreate their tokens, resulting in the new token version.
+4. New tokens are only displayed once at creation time to the user.
+5. New tokens are created following the new format.
+   - A Django migration will be needed to add fields: `hashed_token`, `version`, and `name`.
+   - legacy tokens will be `version = 1` and new tokens `version = 2`
+6. As old tokens are used, they are hashed and stored in the database.
+   - This does not update the token to the new format.
+   - The plaintext token value should be removed in this action.
+
+**In the second release:**
+
+1. As a Django migration, a bulk operation is executed to update all remaining legacy tokens in the database.
+   - This operation will hash the legacy token value, store it in the database, and remove the plaintext value.
+   - It does **not** update the token to the new format.
+
+## Option #2
+
+Instead of slowly generating the hashed token values over time of the legacy tokens (as in Option #1), we could generate a single migration that migrates the entire table to the end state.
+
+### Drawbacks
+
+- This is less than ideal because of the potential performance impact while the migration mutates each row in the table. Potential impacts to self-hosted installs would be unknown as well.
+
+## Option #3
+
+To avoid the two different token versions, we could automatically prepend the prefix `sntyx_` (with `x` just being a placeholder here). We would then follow a similar approach to Option #1 or Option #2 to generate the hashed values.
+
+### Drawbacks
+
+- Users would not be able to benefit from the Github Secrets Scanning since they would still be using their 64 character alphanumeric string without the prefix.
+- Authentication views/logic would become more complex with branching code to handle the with and without prefix cases.
+
+# Unresolved questions
+
+- How do we deploy the necessary changes for the backend and frontend separately and in a backwards compatible way?
+- Implementation path for `APIApplication` secrets.
+- Is there a downside to allowing `null` in the additional columns?
```

Comment:
Every use of the column will need to decide, agree on the semantics of a null value.
A guarantee of meaningful value (i.e. no nulls) will appreciably reduce ambiguity.

Reply:
A `null` is necessary here to keep the migrations safe as far as I can tell.